streams/src/main/java/org/apache/kafka/streams/state/Stores.java

```diff
@@ -0,0 +1,206 @@
+         * @return the key-value store
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context);
+
+        /**
+         * Create a key value store that stores the given type of key and value, that records changes to a Kafka topic, and that
+         * uses the {@link #withTime(Time) configured Time} instance.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keyClass the class for the keys, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @param valueClass the class for the values, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null, or if the {@code keyClass} or {@code valueClass}
+         *             are not one of {@code String.class},
+         *             {@code Integer.class}, {@code Long.class}, or
+         *             {@code byte[].class}
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Class<K> keyClass, Class<V> valueClass);
+
+        /**
+         * Create a key value store that records changes to a Kafka topic and that uses the {@link #withTime(Time) configured
+         * Time} instance and the supplied serializers and deserializers.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keySerializer the serializer for keys; may not be null
+         * @param keyDeserializer the deserializer for keys; may not be null
+         * @param valueSerializer the serializer for values; may not be null
+         * @param valueDeserializer the deserializer for values; may not be null
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Serializer<K> keySerializer, Deserializer<K> keyDeserializer,
+                                                 Serializer<V> valueSerializer, Deserializer<V> valueDeserializer);
+    }
+
+    /**
+     * The factory for creating in-memory key-value stores.
+     */
+    public static interface InMemoryFactory extends KeyValueStoreFactory {
```

Comment:
@guozhangwang. Gotcha. Perhaps the different kind of stores should be first with in-memory vs off-heap at the second level, in case not all types of stores support both. So, for example:

    Stores.keyValue().inMemory().create(...)

Then that brings up several other questions:

1) Should we allow users to create untyped KeyValueStore instances by allowing them to use the ProcessorContext's default serdes? That seems error prone since the default serdes are not typed.

2) So if we remove the ability to implicitly use the default serdes, that allows us to always determine the key and value types based upon parameters. Perhaps we could pass the serdes (or the classes) into the `keyValue` method, and make 'inMemory()' or 'offHeap()' much more of a minor option.

For example, we'd be able to do this:

        KeyValueStore<Integer, String> store1 = Stores.withKeys(Integer.class)
                                                      .withValues(String.class)
                                                      .inMemory()
                                                      .create();

or

        KeyValueStore<Integer, String> store2 = Stores.withKeys(new IntegerSerializer(), new IntegerDeserializer())
                                                      .withValues(new StringSerializer(), new StringDeserializer())
                                                      .inMemory()
                                                      .create();

Do you like this better than the first option I mentioned in this comment?